components/fxa-client/src/commands/send_tab.rs

```diff
@@ -0,0 +1,192 @@
+        let encrypted = Aes128GcmEceWebPushImpl::encrypt(
+            &public_key,
+            &auth_secret,
+            &bytes,
+            WebPushParams::default(),
+        )?;
+        let encrypted = base64::encode_config(&encrypted, base64::URL_SAFE_NO_PAD);
+        Ok(EncryptedSendTabPayload { encrypted })
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct TabData {
+    pub title: String,
+    pub url: String,
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct PrivateSendTabKeys {
+    public_key: Vec<u8>,
+    private_key: Vec<u8>,
+    auth_secret: Vec<u8>,
+}
+
+impl PrivateSendTabKeys {
+    pub fn from_random() -> Result<Self> {
+        let key_pair = LocalKeyPairImpl::generate_random()?;
+        let private_key = key_pair.to_raw();
+        let public_key = key_pair.pub_as_raw()?;
+        let mut auth_secret = vec![0u8; 16];
+        RNG.fill(&mut auth_secret)
+            .map_err(|_| ErrorKind::RngFailure)?;
+        Ok(Self {
+            public_key,
+            private_key,
+            auth_secret,
+        })
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct SendTabKeysPayload {
+    /// Hex encoded kid (kXCS).
+    kid: String,
+    /// Base 64 encoded IV.
+    #[serde(rename = "IV")]
+    iv: String,
+    /// Hex encoded hmac.
+    hmac: String,
+    /// Base 64 encoded ciphertext.
+    ciphertext: String,
+}
+
+impl SendTabKeysPayload {
+    fn decrypt(self, ksync: &[u8], kxcs: &[u8]) -> Result<PublicSendTabKeys> {
+        // Most of the code here is copied from `EncryptedBso::decrypt`:
+        // we can't use that method as-it because `EncryptedBso` forces
+        // a payload id to be specified, which in turns make the Firefox
+        // Desktop commands implementation angry.
+        if hex::decode(self.kid)? != kxcs {
+            return Err(ErrorKind::MismatchedKeys.into());
+        }
+        let key = KeyBundle::from_ksync_bytes(ksync)?;
+        if !key.verify_hmac_string(&self.hmac, &self.ciphertext)? {
+            return Err(ErrorKind::HmacMismatch.into());
+        }
+        let iv = base64::decode(&self.iv)?;
+        let ciphertext = base64::decode(&self.ciphertext)?;
+        let cleartext = key.decrypt(&ciphertext, &iv)?;
+        Ok(serde_json::from_str(&cleartext)?)
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct PublicSendTabKeys {
+    /// URL Safe Base 64 encoded push public key.
+    #[serde(rename = "publicKey")]
+    public_key: String,
+    /// URL Safe Base 64 encoded auth secret.
+    #[serde(rename = "authSecret")]
+    auth_secret: String,
+}
+
+impl PublicSendTabKeys {
+    fn encrypt(&self, ksync: &[u8], kxcs: &[u8]) -> Result<SendTabKeysPayload> {
+        // Most of the code here is copied from `CleartextBso::encrypt`:
+        // we can't use that method as-it because `CleartextBso` forces
+        // a payload id to be specified, which in turns make the Firefox
+        // Desktop commands implementation angry.
```

Comment:
Can we factor out a common helper of some sort, or would that be too awkward? Duplicated crypto code is always kinda scary.

Reply:
Opened https://github.com/mozilla/application-services/pull/700